tensorly/tenalg/svd.py

```diff
@@ -0,0 +1,423 @@
+        Contains the left singular vectors
+
+    Notes
+    -----
+    This function is implemented based on Algorith 5.1 in `Finding structure with randomness:
+    Probabilistic algorithms for constructing approximate matrix decompositions`
+    - Halko et al (2009)
+    """
+    n_eigenvecs, min_dim, max_dim = svd_checks(matrix, n_eigenvecs=n_eigenvecs)
+
+    dim_1, dim_2 = tl.shape(matrix)
+    n_dims = min(n_eigenvecs + n_oversamples, max_dim)
+
+    if (
+        dim_1 > dim_2
+        and n_eigenvecs > min(min_dim, n_dims)
+        or dim_1 < dim_2
+        and n_eigenvecs < min(min_dim, n_dims)
+    ):
+        # transpose matrix to keep the reduced matrix shape minimal
+        matrix_T = tl.transpose(matrix)
+        Q = randomized_range_finder(
+            matrix_T, n_dims=n_dims, n_iter=n_iter, random_state=random_state
+        )
+        Q_H = tl.conj(tl.transpose(Q))
+        matrix_reduced = tl.transpose(tl.dot(Q_H, matrix_T))
+        U, S, V = truncated_svd(matrix_reduced, n_eigenvecs=n_eigenvecs)
+        V = tl.dot(V, tl.transpose(Q))
+    else:
+        Q = randomized_range_finder(
+            matrix, n_dims=n_dims, n_iter=n_iter, random_state=random_state
+        )
+        Q_H = tl.conj(tl.transpose(Q))
+        matrix_reduced = tl.dot(Q_H, matrix)
+        U, S, V = truncated_svd(matrix_reduced, n_eigenvecs=n_eigenvecs)
+        U = tl.dot(Q, U)
+
+    return U, S, V
+
+
+SVD_FUNS = ["truncated_svd", "symeig_svd", "randomized_svd"]
+
+
+def svd_funs(
+    matrix,
+    svd_type="truncated_svd",
+    n_eigenvecs=None,
+    flip_svd=True,
+    non_negative=False,
+    nntype="nndsvd",
+    u_based_decision=True,
+    mask=None,
+    svd_mask_repeats=5,
+    **kwargs,
+):
+    """Dispatching function to various SVD algorithms, alongside additional
+    properties such as resolving sign invariance, imputation, and non-negativity.
+
+    Parameters
+    ----------
+    matrix : tensor
+        A 2D tensor.
+    svd : str, default is 'truncated_svd'
+        Function to use to compute the SVD, acceptable values in tensorly.SVD_FUNS or a callable.
+    n_eigenvecs : int, optional, default is None
+        If specified, number of eigen[vectors-values] to return.
+    flip_svd : bool, optional, default is True
+        Whether to resolve the sign indeterminacy of SVD.
+    non_negative : bool, optional, default is False
+        Whether to make the SVD results non-negative.
+    nntype : str, default is 'nndsvd'
+        Algorithm to use for converting U to be non-negative.
+    u_based_decision : bool, optional, default is True
+        Whether the sign indeterminacy should be resolved using U (vs. V).
+    mask : tensor, default is None.
+        Array of booleans with the same shape as ``matrix``. Should be 0 where
+        the values are missing and 1 everywhere else. None if nothing is missing.
+    svd_mask_repeats : int, default is 5
+        Number of repetitions to apply in missing value imputation.
```

Comment:
That might not be clear to users (though it is if they go through the code..)

Reply:
I tried to at least provide a slightly better name. Eventually I'd like to provide some documentation on handling missing values where this could be explained.

Reply:
Looks good!